File: Tyche.iOS/Bet/Sources/Bet/UI/PoolGamblerBetItemViewModel.swift
import Foundation
import UI
import Core

class PoolGamblerBetItemViewModel: ObservableObject {
    private let betUseCase: BetUseCase
    
    @Published @MainActor private(set) var state: EditableViewState<PoolGamblerBetModel>
    
    @MainActor
    public init(poolGamblerBet: PoolGamblerBetModel, betUseCase: BetUseCase) {
        self.betUseCase = betUseCase
        self.state = .initial(poolGamblerBet)
    }
    
    @MainActor
    func retryBet() {
        if case .failure(_, failed: let poolGamblerBet, _) = state {
            guard let betScore = poolGamblerBet.betScore else { return }
            bet(betScore: betScore)
        }
    }
    
    @MainActor
    func bet(betScore: TeamScore<Int>) {
        Task { [self] in
            var currentPoolGamblerBet = state.value()
            
            let prePoolGamblerBet = currentPoolGamblerBet.copy { builder in
                builder.betScore = betScore
            }
            
            state = .loading(current: currentPoolGamblerBet, target: prePoolGamblerBet)
            
            let bet = Bet(
                poolId: prePoolGamblerBet.poolId,
                gamblerId: prePoolGamblerBet.gamblerId,
                matchId: prePoolGamblerBet.matchId,
                homeTeamBet: BetScore(betScore.homeTeamValue)!,
                awayTeamBet: BetScore(betScore.awayTeamValue)!
            )
            let result = await betUseCase.execute(bet: bet)
            
            switch result {
            case .success(let updatedPoolGamblerBet):
                state = .success(current: currentPoolGamblerBet, succeeded: updatedPoolGamblerBet.toPoolGamblerBetModel())
            case .failure(let error):
                if case BetError.forbidden = error {
                    currentPoolGamblerBet.lock()
                    state = .initial(currentPoolGamblerBet)
                } else {
                    state = .failure(
                        current: currentPoolGamblerBet,
                        failed: prePoolGamblerBet,
                        error: error.toLocalizedError(transformer: { error in error.toBetLocalizedError() }))
                }
            }
        }
    }
    
    @MainActor
    func reset() {
        state = .initial(state.value())
    }
}

File: Tyche.Android/bet/src/main/java/com/felipearpa/tyche/bet/PoolGamblerBetItemViewModel.kt
package com.felipearpa.tyche.bet

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.data.bet.application.BetUseCase
import com.felipearpa.data.bet.domain.Bet
import com.felipearpa.data.bet.domain.BetException
import com.felipearpa.tyche.core.type.BetScore
import com.felipearpa.tyche.core.type.TeamScore
import com.felipearpa.tyche.ui.exception.toLocalizedException
import com.felipearpa.tyche.ui.state.EditableViewState
import com.felipearpa.tyche.ui.state.currentValue
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.flow.updateAndGet
import kotlinx.coroutines.launch
import java.util.UUID

class PoolGamblerBetItemViewModel @AssistedInject constructor(
    @Assisted poolGamblerBet: PoolGamblerBetModel,
    private val betUseCase: BetUseCase
) :
    ViewModel() {

    private val _state =
        MutableStateFlow<EditableViewState<PoolGamblerBetModel>>(
            EditableViewState.Initial(poolGamblerBet)
        )
    val state = _state.asStateFlow()

    fun retryBet() {
        val poolGamblerBet = when (val stateValue = this.state.value) {
            is EditableViewState.Failure -> stateValue.failed
            else -> return
        }
        val betScore = poolGamblerBet.betScore ?: return
        bet(betScore = betScore)
    }

    fun bet(betScore: TeamScore<Int>) {
        viewModelScope.launch {
            val (currentPoolGamblerBet, targetPoolGamblerBet) = _state.updateAndGet { currentState ->
                val currentPoolGamblerBet = currentState.currentValue()
                val targetPoolGamblerBet = currentPoolGamblerBet.copy(betScore = betScore)
                EditableViewState.Loading(
                    current = currentPoolGamblerBet,
                    target = targetPoolGamblerBet
                )
            } as EditableViewState.Loading

            val bet = Bet(
                poolId = currentPoolGamblerBet.poolId,
                gamblerId = currentPoolGamblerBet.gamblerId,
                matchId = currentPoolGamblerBet.matchId,
                homeTeamBet = BetScore(betScore.homeTeamValue),
                awayTeamBet = BetScore(betScore.awayTeamValue)
            )

            betUseCase.execute(bet = bet)
                .onSuccess { updatedPoolGamblerBet ->
                    _state.emit(
                        EditableViewState.Success(
                            old = currentPoolGamblerBet,
                            succeeded = updatedPoolGamblerBet.toPoolGamblerBetModel()
                        )
                    )
                }.onFailure { exception ->
                    if (exception is BetException.Forbidden) {
                        _state.emit(EditableViewState.Initial(currentPoolGamblerBet.copy(isLocked = true)))
                    } else {
                        _state.emit(
                            EditableViewState.Failure(
                                current = currentPoolGamblerBet,
                                failed = targetPoolGamblerBet,
                                exception = exception
                                    .toBetLocalizedExceptionOnMatch()
                                    .toLocalizedException()
                            )
                        )
                    }
                }
        }
    }

    fun reset() {
        _state.update { currentState ->
            val poolGamblerBet = currentState.currentValue()
            EditableViewState.Initial(
                poolGamblerBet.copy(
                    instanceId = UUID.randomUUID().toString()
                )
            )
        }
    }
}